### DISCUSSION TITLE
Best effort zerocopy from Protobuf's ByteString to Netty's ByteBuf

### DISCUSSION THREAD

**Discussion by tisonkun** [netty/netty#13605] [Q&A] [answered]
I'm building a service with gRPC and using Netty's `ByteBuf` as a better byte buffer implementation.

So far, when I receive messages from the gRPC channel, `bytes` field is converted into Protobuf's `ByteString`. My internal service uses `ByteBuf` so there is a necessary conversion:

```java
    public static ByteBuf byteStringToByteBuf(ByteString bytes) {
        final ByteBuffer byteBuffer = bytes.asReadOnlyByteBuffer();
        return Unpooled.buffer(byteBuffer.position()).writeBytes(byteBuffer);
    }
```

I implemented it as above, but this seems always to cause a bytes copy. Are there utilities or libraries to help in doing best-effort zerocopy between these two interfaces? Or is there a better way to write a service instance?

FWIW in my situation the returned bytebuf is always readonly.

**Comment by normanmaurer** [comment, marked as answer]
```
public static ByteBuf byteStringToByteBuf(ByteString bytes) {
    final ByteBuffer byteBuffer = bytes.asReadOnlyByteBuffer();
    return Unpooled.wrappedBuffer(byteBuffer);
}
```

**Reply by tisonkun** [reply]
Thank you! Although, the method name should be `wrappedBuffer`:

```java
    public static ByteBuf byteStringToByteBuf(ByteString bytes) {
        final ByteBuffer byteBuffer = bytes.asReadOnlyByteBuffer();
        return Unpooled.wrappedBuffer(byteBuffer);
    }
```

**Reply by He-Pin** [reply]
Ha, yes, I missunderstand the issue, i was using something like this too.
```scala
  override def write(payload: ByteString): Boolean =
    if (channel.isWritable && channel.isOpen) {
      channel.writeAndFlush(Unpooled.wrappedBuffer(payload.asByteBuffer))
      true
    } else false
```
